### ISSUE TITLE
overwrite protection can be circumvent with renameing

### ISSUE THREAD

**Issue by pseyfert** [jupyter/notebook#1211]
So what I do is the following:

1. start `jupyter notebook`
2. create a notebook "Untitled.ipynb" (python2, but i doubt it matters much)
3. have only one cell with `print "foo"`
4. save it
5. click `File`-> `Close and halt`
6. open the notebook again, twice (such that I have two tabs, each running Untitled.ipynb)
7. change the print statement to `print "bar"` in one of the tabs
8. change the print statement to `print "baz"` in the other tab
9. click the `Save and checkpoint` button in the tab from (7)
10. try and click the `Save and checkpoint` button in the tab from (8)
     (yes, that would overwrite step 9, which is unlikely what I want to do)
Jupyter tells me "The notebook file has changed on disk since the…". Fine, I think "okay, I don't want to throw away the changes from tab 8 (that's why I clicked `save` in the first place) and probably I don't want to throw away the changes on disk (that's why I clicked `save` there). So naively I want to save under a new filename. Yet, I don't have a `save as` button, so I rename the notebook in tab (8) to `Untitled2` and click `save and checkpoint`. No complaint from jupyter.
11. checking the home tab of jupyter, the notebook `Untitled.ipynb` (which I wanted to keep, to have my changes from tab 7, which are not in the version control yet!!!) is gone. I assume it got renamed to `Untitled2` and then overwritten by what was in tab8.

This is with notebook version 4.0.6, jupyter version 1.0.0, ipykernel version 4.3.0, python version 2.7.9

**Comment by minrk** [comment]
Yup, the absence of proper 'Save As' is painful. Rename, which there's not necessarily a reason for you to know, is an operation that moves the file that's already on-disk. In general, there's no decent handling for having one file opened multiple times in the notebook.

If you do find yourself in this situation where you have one file open multiple times and you want to keep both versions, you can save one version, make a copy of it, then write the second version to the original location.

1. <tab1> save
1. <tab1> make a copy (opens <tab3> with a copy of notebook 1)
1. <tab1> close
1. <tab2> save

In your example, you haven't lost any data until you close tab (7), because after renaming in tab (8), you can press save again in tab (7), and the original notebook will be written again.

**Comment by pseyfert** [comment]
Thanks for the instructions. in my real life situation tab (7) is unfortunately long closed and the browser restarted … (actually i observed that my notebook was gone and tried to reproduce how i got rid of it.). But I think i recovered all that got lost.

I guess i then second the request for a save as #1158 

and for my example here. i think it's worth to apply the same check which is called upon 'save' when doing a rename. such that whoever performs the renaming like i did, at least gets warned "hey you're about to move something which some other tab modified in the meanwhile".

**Comment by Phyks** [comment]
> In general, there's no decent handling for having one file opened multiple times in the notebook.

Maybe Jupyter notebook can detect the same file has been opened multiple times and warn the user about it? Or switch to read-only mode, as if the same notebook is opened multiple times it is usually because we want an extra read-only version.

**Comment by minrk** [comment]
There is no way to know that a notebook is open more than once. There is no indication that a notebook "is open," only that it has been opened at some point in the past. The only protection we can have is to tell you that the file has changed since you last opened it when you try to save.

**Comment by Phyks** [comment]
Ok, I was not sure about it. I am just discovering Jupyter's internals, and was thinking that the JS code was maybe periodically polling the webserver.

Not sure if this is something wanted, but I guess it could be achieved with some long polling AJAX query, so that the webserver know the notebook is still open somewhere else.

**Comment by takluyver** [comment]
That would kind of work, but as far as I know there's no good way to distinguish between the user having closed the notebook and the user's internet flaking out, so it may not be all that useful.

**Comment by Phyks** [comment]
Indeed, I was thinking in the setup where notebooks and user were on the same host :/
